src/PostgREST/Middleware.hs

```diff
@@ -39,23 +39,24 @@ runPgLocals :: AppConfig   -> M.HashMap Text JSON.Value ->
                (ApiRequest -> H.Transaction Response) ->
                ApiRequest  -> H.Transaction Response
 runPgLocals conf claims app req = do
-  H.sql $ toS . mconcat $ setSearchPathSql : setRoleSql ++ claimsSql ++ [methodSql, pathSql] ++ headersSql ++ cookiesSql ++ appSettingsSql
-  traverse_ H.sql preReq
+  H.sql . toS $ "select " <> T.intercalate ", " (searchPathSql : roleSql ++ claimsSql ++ [methodSql, pathSql] ++ headersSql ++ cookiesSql ++ appSettingsSql)
+  traverse_ H.sql preReqSql
```

Comment:
What about putting those two statements in a DO block as well?

```
DO LANGUAGE SQL $$
  SELECT set_config(...);
  SELECT preReq();
$$;
```

Should save another round-trip?

Was discussed here: #915

Reply:
You know, I think we can add the preReq after the set_configs, on the same `select`.

I thought the set_configs wouldn't get executed first but looks like they are:
```sql
select 
  set_config('pgrst.a_val', 'look', true), 
  set_config('pgrst.another_val', 'lookie', true), 
  get_guc_value('pgrst.a_val'), 
  get_guc_value('pgrst.another_val');

 set_config | set_config | get_guc_value | get_guc_value
------------+------------+---------------+---------------
 look       | lookie     | look          | lookie
```

WDYT? Do you foresee any issues with that?

Reply:
Hm. Interesting. I could not find anything documented that this execution order is guaranteed. But I was also not able to force a different one. I'm not sure whether we should rely on it.

I thought about doing it in a `WITH`:

```
with set as (
  select
    set_config('pgrst.a_val', 'look', true),
    set_config('pgrst.another_val', 'lookie', true) 
)
select
  get_config('pgrst.a_val'), 
  get_config('pgrst.another_val')
from set;
```

I'd be slightly more confident that the execution order would be as expected here, because of the dependency through `FROM set`. However, even here I could imagine the optimizer doing things differently.

As for my example above: DO with SQL doesn't work, so it would be:

```
DO $$BEGIN
  PERFORM set_config...;
  PERFORM preReq();
END$$;
```

I think one advantage of the DO block would be, that it does not return a result to the client ever, while the plain `SELECT set_config...` would return a row with all the values that were set. But I am not sure whether that is relevant here, when we're not fetching the row?

Reply:
> I'm not sure whether we should rely on it.

I agree. Doesn't look right.

> However, even here I could imagine the optimizer doing things differently.

As long as we use the `WITH q as MATERIALIZED`(default, no need for the MATERIALIZED keyword) we can ensure the cte is executed first, I believe.

Doing the gucs on the`WITH` might be worth exploring. Perhaps we can even include them on our main query somehow.